micropip/package_index.py

```diff
@@ -303,36 +308,101 @@ async def query_package(
 
     index_urls = [PYPI_URL if url == PYPI else url for url in index_urls]
 
-    for url in index_urls:
-        logger.debug("Looping through index urls: %r", url)
-        if _contain_placeholder(url):
-            url = url.format(package_name=name)
-            logger.debug("Formatting url with package name : %r", url)
-        else:
-            url = f"{url}/{name}/"
-            logger.debug("Url has no placeholder, appending package name : %r", url)
-        try:
-            metadata, headers = await fetch_string_and_headers(url, _fetch_kwargs)
-        except HttpStatusError as e:
-            if e.status_code == 404:
-                logger.debug("NotFound (404) for %r, trying next index.", url)
+    projects_info = []
+
+    # With "first-index" strategy, we'll return the first match we find
+    # without checking the other indexes at all.
+    if strategy == "first-index":
+        for url in index_urls:
+            logger.debug("Looping through index urls: %r", url)
+            if _contain_placeholder(url):
+                url = url.format(package_name=name)
+                logger.debug("Formatting url with package name : %r", url)
+            else:
+                url = f"{url}/{name}/"
+                logger.debug("Url has no placeholder, appending package name : %r", url)
+            try:
+                metadata, headers = await fetch_string_and_headers(url, _fetch_kwargs)
+            except HttpStatusError as e:
+                if e.status_code == 404:
+                    logger.debug("NotFound (404) for %r, trying next index.", url)
+                    continue
+                logger.debug(
+                    "Error fetching %r (%s), trying next index.", url, e.status_code
+                )
+                raise
+
+            content_type = headers.get("content-type", "").lower()
+            try:
+                base_url = urlunparse(urlparse(url)._replace(path=""))
+                parser = _select_parser(content_type, name, index_base_url=base_url)
+            except ValueError as e:
+                raise ValueError(f"Error trying to decode url: {url}") from e
+            return parser(metadata)
+    # With "unsafe-first-match" or "unsafe-best-match", we need to check all indexes
+    else:
```

Comment:
I think it is still better to check the strategy string, as people might pass some strange value here. 
Or, we can probably check the value early and failfast.